addons/point_of_sale/models/pos_session.py

```diff
@@ -278,15 +282,468 @@ def _validate_session(self):
+        applied_taxes = self.env['account.tax'].browse(tax_ids)
+        title = 'Sales' if sign == 1 else 'Refund'
+        name = '%s untaxed' % title
+        if applied_taxes:
+            name = '%s with %s' % (title, ', '.join([tax.name for tax in applied_taxes]))
+        base_tags = applied_taxes\
+            .mapped('invoice_repartition_line_ids' if sign == 1 else 'refund_repartition_line_ids')\
+            .filtered(lambda line: line.repartition_type == 'base')\
+            .tag_ids
+        partial_vals = {
+            'name': name,
+            'account_id': account_id,
+            'move_id': self.move_id.id,
+            'tax_ids': [(6, 0, tax_ids)],
+            'tag_ids': [(6, 0, base_tags.ids)],
+        }
+        return self._credit_amounts(partial_vals, amount, amount_converted)
+
+    def _get_tax_vals(self, key, amount, amount_converted, base_amount):
+        account_id, repartition_line_id, tax_id, tag_ids = key
+        tax = self.env['account.tax'].browse(tax_id)
+        partial_args = {
+            'name': tax.name,
+            'account_id': account_id,
+            'move_id': self.move_id.id,
+            'tax_base_amount': base_amount,
+            'tax_repartition_line_id': repartition_line_id,
+            'tag_ids': [(6, 0, tag_ids)],
+        }
+        return self._credit_amounts(partial_args, amount, amount_converted)
+
+    def _get_stock_expense_vals(self, exp_account, amount, amount_converted):
+        partial_args = {'account_id': exp_account.id, 'move_id': self.move_id.id}
+        return self._debit_amounts(partial_args, amount, amount_converted)
+
+    def _get_stock_output_vals(self, out_account, amount, amount_converted):
+        partial_args = {'account_id': out_account.id, 'move_id': self.move_id.id}
+        return self._credit_amounts(partial_args, amount, amount_converted)
+
+    def _get_statement_line_vals(self, statement, receivable_account, amount):
+        return {
+            'date': fields.Date.context_today(self),
+            'amount': amount,
+            'name': self.name,
+            'statement_id': statement.id,
+            'account_id': receivable_account.id,
+        }
+
+    def _update_amounts(self, old_amounts, amounts_to_add, date, round=True):
+        new_amounts = {}
+        for k, amount in old_amounts.items():
+            if k == 'amount_converted':
+                amount_converted = old_amounts['amount_converted']
+                amount_to_convert = amounts_to_add['amount']
+                new_amounts['amount_converted'] = amount_converted if self.is_in_company_currency else (amount_converted + self._amount_converter(amount_to_convert, date, round))
+            else:
+                new_amounts[k] = old_amounts[k] + amounts_to_add[k]
+        return new_amounts
+
+    def _round_amounts(self, amounts):
+        new_amounts = {}
+        for key, amount in amounts.items():
+            new_amounts[key] = self.currency_id.round(amount)
```

Comment:
maybe the currency used to round should be either self.currency_id or self.company_id.currency_id depending if we are rounding the amount or the amount_converted.